renderdoc/driver/vulkan/vk_manager.cpp

```diff
@@ -682,3 +682,40 @@ bool VulkanResourceManager::ResourceTypeRelease(WrappedVkRes *res)
 {
   return m_Core->ReleaseResource(res);
 }
+
+void VulkanResourceManager::Serialise_DeviceMemoryRefs(WriteSerialiser &ser)
```

Comment:
This needs a couple of changes to follow the standard pattern for `Serialise` functions:

* Serialise functions shouldn't call `SCOPED_SERIALISE_CHUNK` internally, that macro is intended for the external function to set up and call into the Serialise function for the implementation. So in this case you'd have a function named something like `InsertDeviceMemoryRefs` which would do that.
* Serialise functions should be written as reflexive - so it will conditionally do any capture-time setup (if needed), serialise the stored data, then do any replay-time setup. So for example it would look like:
  ```
  template<typename SerialiserType>
  bool VulkanResourceManager::Serialise_DeviceMemoryRefs(SerialiserType &ser,
                                                         const std::vector<HalfInterval> &data)
  {
    SERIALISE_ELEMENT(data);

    SERIALISE_CHECK_READ_ERRORS();

    if(IsReplayingAndReading())
    {
      // unpack data into m_MemFrameRefs
    }

    return true;
  }

  void VulkanResourceManager::InsertDeviceMemoryRefs(WriteSerialiser &ser)
  {
    std::vector<HalfInterval> data;

    // pack m_MemFrameRefs into data here

    {
      SCOPED_SERIALISE_CHUNK(SystemChunk::InitialContents, sizeEstimate);
      Serialise_DeviceMemoryRefs(ser, data);
    }
  }
  ```

  And the chunk dispatch in `ProcessChunk` can do `return Serialise_DeviceMemoryRefs(ser, {});`

I know `Serialise_InitialContentsNeeded` / `CreateInitialContents` break this pattern and honestly it should be refactored. `InsertInitialContentsChunks` is probably a better template. It's more of a historical oddity and in this case all we are doing is saving and restoring `m_MemFrameRefs` so we should be able to easily follow the standard pattern.

I'd also slightly prefer having just one chunk with this data rather than a chunk per ID. If there's a strong reason to keep it as a chunk-per-ID that's fine, e.g. if you're thinking of stripping out elements that aren't used in the frame. That seems like it's not necessary though and I feel it will be simpler to serialise `m_MemFrameRefs` in one block.

You could either serialise a sorted vector of three-tuples with ID, start, value and then reconstruct `m_MemFrameRefs` from that, or you could serialise a vector of `std::pair<ResourceId, MemRefs>` and then define the serialisation for `MemRefs` to serialise the ID and a vector of `HalfInterval` type objects. It depends on what you think would be cleaner.

Reply:
This all makes sense.

The one piece that feels kind of messy here is all the extra copies to convert between maps and vectors. What do you think of adding a `std::map` serialiser (based on the `std::vector` serialiser in [serialiser.h](https://github.com/baldurk/renderdoc/blob/v1.x/renderdoc/serialise/serialiser.h#L409)? This would make the serialisation of `m_MemFrameRefs` trivial, without any extra copying.

That being said, I doubt the overhead of the copies would actually be noticeable, so if you prefer not to have `std::map` serialisation, then I'll just implement it as you suggested.

Reply:
I definitely agree that it's not pretty. I did think about this but I'm not entirely sure where that path leads. For one thing the serialisation is closely tied to the structured data representation. Currently a list is hand-waved away as just a plain array type which is probably fine, but a map feels less similar. It could be represented as a 'struct' with two parallel arrays: keys and values, or as an array of pairs, or maybe even the structured data should have a native notion of a key-value store.

In the end I think as you say it's probably not going to be noticeable in terms of perf so it's probably not worth opening that can of worms. If we do decide to serialise maps natively in future then it won't be too bad to change over the serialisation and version it for backwards compatibility.